src/dp/distributions.rs

```diff
@@ -0,0 +1,340 @@
+/// using the given random number generator for base randomness.
+fn sample_bernoulli_exp<R: Rng + ?Sized>(gamma: &Ratio<BigUint>, rng: &mut R) -> bool {
+    assert!(!gamma.denom().is_zero());
+    // sample floor(n/d) independent Bernoulli(exp(-1))
+    // If all are 1, return Bernoulli(exp(-(gamma-floor(gamma))))
+    let mut gamma: Ratio<BigUint> = gamma.clone();
+    while Ratio::<BigUint>::one() < gamma {
+        if !sample_bernoulli_exp1(&Ratio::<BigUint>::one(), rng) {
+            return false;
+        }
+        gamma -= Ratio::<BigUint>::one();
+    }
+    sample_bernoulli_exp1(&gamma, rng)
+}
+
+/// sample from the geometric distribution with parameter 1 - exp(-gamma) (slow).
+///
+/// `sample_geometric_exp_slow(gamma, rng)` returns numbers distributed according to
+/// $Geometric(1 - exp(-gamma))$, using the given random number generator for base randomness.
+fn sample_geometric_exp_slow<R: Rng + ?Sized>(gamma: &Ratio<BigUint>, rng: &mut R) -> BigUint {
+    assert!(!gamma.denom().is_zero());
+    let mut k = BigUint::zero();
+    loop {
+        if sample_bernoulli_exp(gamma, rng) {
+            k += 1u8;
+        } else {
+            return k;
+        }
+    }
+}
+
+/// sample from the geometric distribution  with parameter 1 - exp(-gamma) (fast).
+///
+/// `sample_geometric_exp_fast(gamma, rng)` returns numbers distributed according to
+/// $Geometric(1 - exp(-gamma))$, using the given random number generator for base randomness.
+fn sample_geometric_exp_fast<R: Rng + ?Sized>(gamma: &Ratio<BigUint>, rng: &mut R) -> BigUint {
+    let d = gamma.denom();
+    assert!(!d.is_zero());
+    if gamma.is_zero() {
+        return BigUint::zero();
+    }
+
+    // sample uniform biguint in [0,d)
+    let usampler = Uniform::new(BigUint::zero(), d);
+    let mut u = usampler.sample(rng);
+
+    while !sample_bernoulli_exp(&Ratio::<BigUint>::new(u.clone(), d.clone()), rng) {
+        u = usampler.sample(rng);
+    }
+
+    let v2 = sample_geometric_exp_slow(&Ratio::<BigUint>::one(), rng);
+    v2 * d + u / gamma.numer()
+}
+
+/// sample from the discrete laplace distribution.
+///
+/// `sample_discrete_laplace(scale, rng)` returns numbers distributed according to
+/// $\mathcal{L}_\mathbb{Z}(0, scale)$, using the given random number generator for base randomness.
+///
+/// # Citation
+/// * [CKS20 The Discrete Gaussian for Differential Privacy](https://arxiv.org/abs/2004.00010)
+fn sample_discrete_laplace<R: Rng + ?Sized>(scale: &Ratio<BigUint>, rng: &mut R) -> BigInt {
+    let (n, d) = (scale.numer(), scale.denom());
+    assert!(!d.is_zero());
+    if n.is_zero() {
+        return BigInt::zero();
+    }
+
+    loop {
+        let negative = sample_bernoulli_standard(rng);
+        let magnitude: BigInt = sample_geometric_exp_fast(&scale.recip(), rng).into();
+        if negative || !magnitude.is_zero() {
+            return if negative { -magnitude } else { magnitude };
+        }
```

Comment:
the pseudocode and python implementation say to return `-Y` (i.e. `-magnitude`) in case `B=1` (i.e. our `negative=true`, or in the opendp implementation `positive=true`). so i think `if negative { -magnitude } else { magnitude }` is the correct way?

Reply:
What does the reference implementation say? Let's follow the reference implementation and add a comment to note the discrepancy with the paper.

Reply:
there is no discrepancy between the python reference impl and the paper at this particular point. there is a difference to the opendp implementation. i stuck to the python/paper version.